config/config.go

```diff
@@ -198,6 +203,130 @@ type HTTPRewriteResponses struct {
+//
+// Returns an error if there is an applicable rule which does not find a valid client certificate subject.
+func VerifyClientCertMetadata(rules []VerifyClientCertificateMetadataRule, chains [][]*x509.Certificate, logger logger.Logger) error {
+	for _, rule := range rules {
+		for _, chain := range chains {
+			requiredSubject := rule.CASubject.ToName()
+			if !checkIfRuleAppliesToChain(chain, logger, requiredSubject) {
+				continue
+			}
+			err := checkClientCertificateMetadataRule(chain, logger, rule)
+			if err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+// checkIfRuleAppliesToChain checks, whether the provided certificate chain contains a CA certificate
+// whose subject matches the requiredCASubject name.
+//
+// Returns true, if a CA certificate in the chain (cert.IsCA == true) matches the requiredCASubject name.
+func checkIfRuleAppliesToChain(chain []*x509.Certificate, logger logger.Logger, requiredCASubject pkix.Name) bool {
+	for i, cert := range chain {
+		logger.Debug("cert", zap.Int("index", i), zap.Bool("ca", cert.IsCA), zap.String("subject", cert.Subject.String()), zap.String("issuer", cert.Issuer.String()))
+		if cert.IsCA && requiredCASubject.ToRDNSequence().String() == cert.Subject.ToRDNSequence().String() {
+			return true
+		}
+	}
+	return false
+}
+
+// checkClientCertificateMetadataRule is called by checkIfRuleAppliesToChain. When the CA subject matches, the subject
+// of the client certificate is compared agains the subjects defined in rule.
+//
+// Returns an error when:
+// * the certificate does not match any of the ValidSubjects in rule.
+// * the chain does not contain a client certificate (i.e. IsCA == false).
+func checkClientCertificateMetadataRule(chain []*x509.Certificate, logger logger.Logger, rule VerifyClientCertificateMetadataRule) error {
+	for _, cert := range chain {
+		if !cert.IsCA {
+			subject := cert.Subject
+			for _, validSubject := range rule.ValidSubjects {
+				vaildCertSubject := validSubject.ToName()
+				if vaildCertSubject.ToRDNSequence().String() == subject.ToRDNSequence().String() {
+					logger.Debug("chain", zap.String("issuer", cert.Issuer.String()), zap.Bool("CA", cert.IsCA), zap.String("subject", cert.Subject.String()))
+					return nil
+				}
+			}
+			logger.Warn("invalid-subject", zap.String("issuer", cert.Issuer.String()), zap.String("subject", cert.Subject.String()), zap.Object("allowed", rule.ValidSubjects))
+			return fmt.Errorf("subject not in the list of allowed subjects for CA Subject %q: %q", rule.CASubject, subject)
+		}
+	}
+	return fmt.Errorf("incomplete chain")
+}
+
+// InitClientCertMetadataRules compares the defined rules against client CAs set in `client_ca_certs`. When a rule
+// is found that does not have a corresponding client CA (based on the CA's subject) that matches the rule, startup will fail.
+//
+// This is to avoid defining a rule with a minor typo that would then not apply at all and would make the whole
```

Comment:
This makes sense that we would want to check for typos, but I worry that failing at runtime is inconvenient and late in the process. Is there a way to fail at templating time instead (or in addition)?

Reply:
Yes, at least in addition. While it's unlikely that e.g. Gorouter would be used outside the routing-release, I would prefer to have a safeguard in the Gorouter code anyway.

Reply:
in addition is great!